### PR TITLE
commitlog: Ensure we never have more than one new_segment call at a time

### PR DESCRIPTION
**Pull request by elcallio**
Refs #9896

Found by @eliransin. Call to new_segment was wrapped in with_timeout.
This means that if primary caller timed out, we would leave new_segment
calls running, but potentially issue new ones for next caller.

This could lead to reserve segment queue being read simultanously. And
it is not what we want.

Change to always use the shared_future wait, all callers, and clear it
only on result (exception or segment)

(backport to 4.6)

### CODE REVIEW

File: db/commitlog/commitlog.cc [hunk 1]
```diff
@@ -1586,6 +1586,8 @@ future<db::commitlog::segment_manager::sseg_ptr> db::commitlog::segment_manager:
 }
 
 future<db::commitlog::segment_manager::sseg_ptr> db::commitlog::segment_manager::new_segment() {
+    gate::holder g(_gate);
+
```
Comment (bhalevy) [inline db/commitlog/commitlog.cc:1590]: In `shutdown()` we both set `_shutdown` and close `_gate` so if the gate is closed here we won't get to the statement below.
Why not hold the gate after checking `_shutdown`?

**Review by bhalevy — CHANGES_REQUESTED** [review]
Other than the placement of gate::holder LGTM

File: db/commitlog/commitlog.cc (hunk 1 above)
Comment (elcallio) [reply db/commitlog/commitlog.cc:1590]: Just same order as old code. Either is fine I guess...

Comment (bhalevy) [reply db/commitlog/commitlog.cc:1590]: I'd rather not change existing errors as tests may depend on them.

Comment (elcallio) [reply db/commitlog/commitlog.cc:1590]: Then this is correct, as it is the order we took things before (gate was held in calling method).

Comment (bhalevy) [reply db/commitlog/commitlog.cc:1590]: Ah, I see

**Review by bhalevy — APPROVED** [review]

**Comment by nyh** [comment]
> Refs #9896
> 
> Found by @eliransin. Call to new_segment was wrapped in with_timeout. This means that if primary caller timed out, we would leave new_segment calls running, but potentially issue new ones for next caller.
> 
> This could lead to reserve segment queue being read simultanously. And it is not what we want.
> 
> Change to always use the shared_future wait, all callers, and clear it only on result (exception or segment)
> 
> (backport to 4.6)

This "backport to 4.6" message only appears here, but won't appear in the git log when the single-patch commit is done (and doesn't include this message).
Also, the fact you only "Refs" and not "Fixes" #9896 will be another reason we'll forget to backport it (if it fixes an issue, we'll go back to it and remember we need to backport).

**Review by nyh — APPROVED** [review]

**Review by eliransin — APPROVED** [review]
LGTM :+1:

**Comment by eliransin** [comment]
@avikivity this is the PR we discussed that should fix the segfault on 4.6
(single consumer violation in queue).

**Comment by slivne** [comment]
@Avi Kivity ***@***.***> / @Nadav Har'El ***@***.***>
Backport to 4.6 please - this is blocking the release

**Comment by avikivity** [comment]
Backported to 4.6. Re further backports, are they needed? I see with_timeout was added in 2016 (31645e2c4a61184c5884bc28551c1ad1eeacb7da), and that's the source of the problem since it lets the original allocation continue in the background after timeout. However, it may happen to be safe due to reference counting.

**Comment by eliransin** [comment]
@avikivity I will check this,  I have a way to check this with a simple code modification and a unit test run (assuming the CL code is similar in 4.4 and 4.5).
Will check back with results soon.

**Comment by elcallio** [comment]
No, 4.5 is safe. It is still on non-coroutine version.

**Comment by avikivity** [comment]
Why is it safe? coroutines by themselves don't cause the issue, do they? I understood it's with_timeout() (which should probably be deprecated) that is the root cause.

**Comment by elcallio** [comment]
The 4.5 code does not use `with_timeout`, only `shared_future::get_future(timeout)`, so it does not forget background tasks. Like what patch above reverted back to. Sloppy coroutine rewrite on my part.

**Comment by avikivity** [comment]
Aha.

/cc @gleb-cloudius - we talked about with_timeout() being dangerous.

**Comment by gleb-cloudius** [comment]
On Tue, Feb 01, 2022 at 11:49:53AM -0800, Avi Kivity wrote:
> Aha.
> 
> /cc @gleb-cloudius - we talked about with_timeout() being dangerous.
> 
That is no way a with_timeout() fault here. Every tool can be misused.
Here the code wanted to serialize an access to a resource, but allowed
background task to access the resource without serialization. Your
proposed abort_target interface can be abused in exactly same way. Bugs
happen.

--
			Gleb.

**Comment by eliransin** [comment]
@gleb-cloudius , maybe we could have somehow included a reference to the original future / task in the timeout exception (or some derived class), this will force the user to do something about the original task like pinning it to something.
There is no argument that the tool as it is was misused, however, it looks like abandoning tasks that are no longer interesting (obsoleted by timeout) is an anti pattern.

**Comment by gleb-cloudius** [comment]
On Wed, Feb 02, 2022 at 01:18:21AM -0800, Eliran Sinvani wrote:
> @gleb-cloudius , maybe we could have somehow included a reference to the original future / task in the timeout exception (or some derived class), this will force the user to do something about the original task like pinning it to something.
How would that helped here exactly?

> There is no argument that the tool as it is was misused, however, it looks like abandoning tasks that are no longer interesting (obsoleted by timeout) is an anti pattern.
> 
The problem here was not that nobody waits for the background task, but
that the code failed to synchronise with the background task. Its like
writing in posix:

sem_wait(sem, timeout);
// access resource without checking for timeout.

In the ideal world you do not have a background task after timeout at all
(nobody needs its result anyway) and with_timeout() is only a "crutch"
for tasks that do not obey the timeout. But when you use it you have to
realize that there is still a task in the background.

--
			Gleb.

**Pull request closed without merging**